Add code organization section and update feature flag & permission gu…

rsund55 opened the pull request:
Updating our feature flag and permission guides to reflect changes to our processes. Also adding a new `code-design` section to start building some guides around our use of `packwerk`. 

Updates include:

- Added details about permission descriptions and adding new permissions to the role & permission allow list
- Updates to the feature flag guides, use them by default, examples on when and when not to use them and using data migrations when adding new ones.
- Added new `code-design` section with a few packwerk guides.

tute reviewed (approved):
👌

craineum commented on authorization/README.md:
```diff
@@ -7,7 +7,11 @@
 - Don't update roles or permissions directly within Auth0.
 - Use Terraform in the [DevOps] repository to update permissions.
-- Ensure the [Permissions Matrix] is up to date.
+- Ensure new permissions have detailed descriptions. Our customers need to understand
```
Should we state where to document this?

rsund55 replied:
You mean something like `Ensure new permissions have detailed descriptions in terraform.`

craineum commented on code-organization/README.md:
```diff
@@ -0,0 +1,12 @@
+
+- Prefer creating packs to encompass business flows (i.e. Processing, Payments).
+- Prefer adding pack specific models and concerns inside the pack if possible.
+- Prefer setting stewards.
```
Should we state where this happens and what it should look like?

craineum commented on feature-flags/README.md:
```diff
@@ -17,7 +17,16 @@ of 3 categories:
+    - SOPs are impacted
+    - Training required
+    - Release flexibility
+  - Examples where feature flags may not be needed:
```
Could we bold "may not" to emphasis that these aren't always true.

craineum commented on feature-flags/README.md:
```diff
@@ -17,7 +17,16 @@ of 3 categories:
+    - Training required
+    - Release flexibility
+  - Examples where feature flags may not be needed:
+    - Label changes
```
Not sure I would include this as many of the label changes can impact SOPs, and it is hard to know what is in our customer SOPs.

rsund55 replied:
Thanks for the feedback. Do you have any suggestions for other examples of when a feature flag may not be needed? I was trying to provide a few real world examples to add clarity.

craineum commented on code-organization/README.md:
```diff
@@ -0,0 +1,12 @@
+# Packwerk
```
I like that you title the page `code-organization`, but would carry that to this title, then have a "Packwerk" section.

The reason I really like this is because it sparked an idea of documenting how we use "Form Models" or "Service Models" (or maybe they are separate things?) to organize our code over using "Service Objects".

rayfaddis replied:
I like where you are going @craineum  and think we possibly need more of that documented. 

We do have a section already with a few OO things here https://github.com/BuoySoftware/guides/tree/1cd099841ca46aeae3aa25d79b9d0154c762db3a/object-oriented-design

So I actually wonder if the OOD pages bullets should be moved here, or vice versa, if we should be putting a Packwerk section in the OOD page and later adding more around what else you have mentioned?

rsund55 replied:
Renamed `code-organization` to `code-design`. Deleted the separate OOD page and added it to the `code-design` page instead.

craineum reviewed (approved):
Looks great! Thanks for working on this.

rsund55 commented:
@rayfaddis @craineum Cleaned this up a bit based on your feedback, take another look and let me know what you think. The highlight is that I renamed `code-organization` to `code-design` and added the OOD guides in there and removed the old OOD page.

rayfaddis commented on code-design/README.md:
```diff
@@ -1,10 +1,25 @@
+At Buoy we utilize [packwerk] to enable us to organize our code into
+packages based on domain concepts or business flows.
+
+### Guide
```
We had mentioned an `avoid` here, wasn't it around making team based packs?

rsund55 replied:
Done

rsund55 commented on code-design/README.md:
```diff
@@ -1,10 +1,25 @@
 - Limit dependencies of an object (entities an object depends on).
 - Limit an object's dependents (entities that depend on an object).
-- [Prefer composition over inheritance.]
+- [Prefer composition over inheritance]
```
Done
